### PR TITLE
Heartbeat and graceful termination

### LINKED ISSUE #4: Panic on reload
апр 12 00:26:16 vm-moira moira-cache[6466]: panic: runtime error: invalid memory address or nil pointer dereference
апр 12 00:26:16 vm-moira moira-cache[6466]: [signal 0xb code=0x1 addr=0x0 pc=0x51ad90]
апр 12 00:26:16 vm-moira moira-cache[6466]: goroutine 26 [running]:
апр 12 00:26:16 vm-moira moira-cache[6466]: github.com/moira-alert/cache/filter.findPart(0xc8200c9100, 0x3, 0x39, 0xc820142c20, 0x1, 0x1, 0x0, 0x0, 0x0, 0x533e23)
апр 12 00:26:16 vm-moira moira-cache[6466]: /opt/TeamCity/buildAgent/work/8f0b3fe99b08b4df/gocode/src/github.com/moira-alert/cache/filter/pattern.go:151 +0x450
апр 12 00:26:16 vm-moira moira-cache[6466]: github.com/moira-alert/cache/filter.(*PatternStorage).MatchPattern(0xc8200c00a8, 0xc8200c9100, 0x27, 0x39, 0x0, 0x0, 0x0)
апр 12 00:26:16 vm-moira moira-cache[6466]: /opt/TeamCity/buildAgent/work/8f0b3fe99b08b4df/gocode/src/github.com/moira-alert/cache/filter/pattern.go:124 +0x1c2
апр 12 00:26:16 vm-moira moira-cache[6466]: github.com/moira-alert/cache/filter.(*PatternStorage).ProcessIncomingMetric(0xc8200c00a8, 0xc8200c9100, 0x38, 0x39, 0x0)
апр 12 00:26:16 vm-moira moira-cache[6466]: /opt/TeamCity/buildAgent/work/8f0b3fe99b08b4df/gocode/src/github.com/moira-alert/cache/filter/metric.go:98 +0x7d0
апр 12 00:26:16 vm-moira moira-cache[6466]: main.handleConnection.func1(0xc820188000, 0xc8200c9100, 0x39, 0x39, 0xc82017a000)
апр 12 00:26:16 vm-moira moira-cache[6466]: /opt/TeamCity/buildAgent/work/8f0b3fe99b08b4df/gocode/src/github.com/moira-alert/cache/main.go:198 +0x6c
апр 12 00:26:16 vm-moira moira-cache[6466]: created by main.handleConnection
апр 12 00:26:16 vm-moira moira-cache[6466]: /opt/TeamCity/buildAgent/work/8f0b3fe99b08b4df/gocode/src/github.com/moira-alert/cache/main.go:201 +0x300

### PR DESCRIPTION
**Pull request by gmlexx**
Refactoring of selfcheck.
Try fix #4 and #6 
Can not reproduce this crushes, but hope grafefull termination will help.

### CODE REVIEW

File: filter/db.go [hunk 1]
```diff
@@ -30,17 +30,18 @@ func GetMetricRetentionDbKey(metric string) string {
...
+	err := c.Send("INCR", "moira-selfstate:metrics-heartbeat")
+	return err
 }
 
-func (connector *DbConnector) selfStateSave(tc int64) error {
+func (connector *DbConnector) getPatterns() ([]string, error) {
 	c := connector.Pool.Get()
 	defer c.Close()
-	err := c.Send("SET", "moira-selfstate:last-metric-received-ts", fmt.Sprintf("%v", tc), "EX", fmt.Sprintf("%v", 60))
-	return err
```
Comment (beevee) [inline filter/db.go]: Stupid diff, took several minutes to understand what's happening here :)

File: heartbeat.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+)
+
+func heartbeat(db *filter.DbConnector, terminate chan bool, wg *sync.WaitGroup) {
+	defer wg.Done()
+	count := filter.TotalMetricsReceived.Count()
+	for {
+		select {
+		case <-terminate:
+			return
+		case <-time.After(time.Second * 5):
```
Comment (beevee) [inline heartbeat.go]: Why do we need to check heartbeat that often? In case of low load metrics can arrive once every minute.

Comment (gmlexx) [inline heartbeat.go]: This is not a check, this is heartbeat sending.
5 sec should be enough for any case and does't hit a performance I guess.
If we introduce new settings, user can misconfigure it.

**Comment by beevee** [comment]
OK to merge, but waiting for changes in notifier.

**Comment by AlexAkulov** [comment]
Cache failed 
in log 
```
pid:23789 21:55:40.411871 main.go:108: listening on :2003
```
stdout 
```
Jun 01 21:47:40 vm-edi-graph3.kontur systemd[1]: Started moira-cache - metric stream filtering and caching for Moira.
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: panic: runtime error: invalid memory address or nil pointer dereference
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: [signal 0xb code=0x1 addr=0x0 pc=0x4834b0]
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: goroutine 128 [running]:
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: panic(0x67ff20, 0xc8200100a0)
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: /usr/local/go/src/runtime/panic.go:464 +0x3e6
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: github.com/moira-alert/cache/filter.findPart(0xc820126b40, 0x3, 0x8b, 0xc82002cc10, 0x1, 0x1, 0x0, 0x0, 0x0, 0x554533)
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: /usr/share/gocode/src/github.com/moira-alert/cache/filter/pattern.go:157 +0x450
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: github.com/moira-alert/cache/filter.(*PatternStorage).MatchPattern(0xc82001e0c0, 0xc820126b40, 0x7a, 0x8b, 0x0, 0x0, 0x0)
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: /usr/share/gocode/src/github.com/moira-alert/cache/filter/pattern.go:130 +0x1c2
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: github.com/moira-alert/cache/filter.(*PatternStorage).ProcessIncomingMetric(0xc82001e0c0, 0xc820126b40, 0x8a, 0x8b, 0x0)
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: /usr/share/gocode/src/github.com/moira-alert/cache/filter/metric.go:96 +0x84a
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: main.handleConnection.func2(0xc8200b6c00, 0xc820126b40, 0x8b, 0x8b, 0xc820055ec0)
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: /usr/share/gocode/src/github.com/moira-alert/cache/main.go:215 +0x6c
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: created by main.handleConnection
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: /usr/share/gocode/src/github.com/moira-alert/cache/main.go:218 +0x383
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: panic: runtime error: invalid memory address or nil pointer dereference
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: [signal 0xb code=0x1 addr=0x0 pc=0x4834b0]
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: goroutine 72 [running]:
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: panic(0x67ff20, 0xc8200100a0)
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: /usr/local/go/src/runtime/panic.go:464 +0x3e6
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: github.com/moira-alert/cache/filter.findPart(0xc82007b380, 0x3, 0x7c, 0xc820144c10, 0x1, 0x1, 0x0, 0x0, 0x0, 0x554533)
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: /usr/share/gocode/src/github.com/moira-alert/cache/filter/pattern.go:157 +0x450
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: github.com/moira-alert/cache/filter.(*PatternStorage).MatchPattern(0xc82001e0c0, 0xc82007b380, 0x6b, 0x7c, 0x0, 0x0, 0x0)
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: /usr/share/gocode/src/github.com/moira-alert/cache/filter/pattern.go:130 +0x1c2
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: github.com/moira-alert/cache/filter.(*PatternStorage).ProcessIncomingMetric(0xc82001e0c0, 0xc82007b380, 0x7b, 0x7c, 0x0)
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: /usr/share/gocode/src/github.com/moira-alert/cache/filter/metric.go:96 +0x84a
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: main.handleConnection.func2(0xc8200b6c00, 0xc82007b380, 0x7c, 0x7c, 0xc820055ec0)
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: /usr/share/gocode/src/github.com/moira-alert/cache/main.go:215 +0x6c
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: created by main.handleConnection
Jun 01 21:47:40 vm-edi-graph3.kontur moira-cache[20791]: /usr/share/gocode/src/github.com/moira-alert/cache/main.go:218 +0x383
Jun 01 21:47:40 vm-edi-graph3.kontur systemd[1]: moira-cache.service: main process exited, code=exited, status=2/INVALIDARGUMENT
Jun 01 21:47:40 vm-edi-graph3.kontur systemd[1]: Unit moira-cache.service entered failed state.
```

**Comment by gmlexx** [comment]
Fix in @28af065

**Pull request merged**